check/lib/Util.hs

```diff
@@ -0,0 +1,136 @@
+import System.Process
+import qualified Text.Parsec as P
+import qualified Text.Parsec.String as P
+
+inCabalNewRepl :: String -> String -> IO String
+inCabalNewRepl module_ input_ = do
+    -- if HC variable is set, we use it to invoke 'cabal new-repl'
+    envvars <- getEnvironment
+    let mhc = lookup "HC" envvars
+    let withCompiler = maybe [] (\hc -> ["-w", hc]) mhc
+
+    putStrLn "=== INPUT ==="
+    putStrLn input
+    output <- readProcess "cabal" (args withCompiler) input
+    putStrLn "=== OUTPUT ==="
+    putStrLn output
+
+    return output
+  where
+    args withCompiler =
+        [ "new-repl"
+        , "for-repl"
+        , "--ghc-option=-ignore-dot-ghci"
+        , "--enable-tests"
+        , "--enable-benchmarks"
+        ] ++ withCompiler ++
+        [ "--"
+        ]
+
+    input = unlines
+        -- load module
+        [ ":set prompt \"\""
+        , ":l " ++ module_
+        , show randomString1
+        ]
+        ++ input_ ++
+        unlines
+            [ "", show randomString2
+            ]
+
+-- markers
+randomString1 :: String
+randomString1 = "XHyIGBiBHqPo7PdVNIJV2yAwh60czbwFVKM8FhCL"
+
+randomString2 :: String
+randomString2 = "bo1vzYNeOweLxLi2zlLjbHKjXT8nClvdKKSQaehs"
+
+normalizeLines :: String -> String
+normalizeLines
+    = unlines
+    . joinIndentedLines Nothing
+    . lines
+  where
+    joinIndentedLines :: Maybe String ->  [String] -> [String]
+    joinIndentedLines Nothing  []  = []
+    joinIndentedLines (Just l) [] = [l]
+    joinIndentedLines Nothing  (l : ls) = joinIndentedLines (Just l) ls
+    joinIndentedLines (Just p) ((' ':' ':l) : ls) =
+        joinIndentedLines (Just $ p ++ ' ' : l) ls
+    joinIndentedLines (Just p) (l : ls) =
+        p : joinIndentedLines (Just l) ls
+
+normalizeSignatures :: String -> String
+normalizeSignatures
+    = unlines
+    . map stripModuleIdentifiers   -- strip package-prefixes from types
+    . takeWhile (/= show randomString2)
+    . drop 1
+    . dropWhile (/= show randomString1)
+    . joinIndentedLines Nothing
+    . lines
+  where
+    --
+    -- @
+    -- foo
+    --    :: a
+    --    -> b
+    -- @
+    --
+    -- into
+    --
+    -- @
+    -- foo :: a -> b
+    -- @
+    joinIndentedLines :: Maybe String ->  [String] -> [String]
+    joinIndentedLines Nothing  []  = []
+    joinIndentedLines (Just l) [] = [l]
+    joinIndentedLines Nothing  (l : ls) = joinIndentedLines (Just l) ls
+    joinIndentedLines (Just p) ((' ':' ':l) : ls) =
+        joinIndentedLines (Just $ p ++ ' ' : l) ls
+    joinIndentedLines (Just p) (l : ls) =
+        p : joinIndentedLines (Just l) ls
+
+    -- regex is better than manual parsing
+    --
+    -- replace
+    --
+    -- @semigroup-version@tag:C@ with @C"
+    --
+    -- so `typediff` can parse the types
+    --
+    stripModuleIdentifiers = replace parser
+      where
+        parser = do
+            -- packages
+            _ <- P.choice
+                [ P.try (P.string "ghc-prim")
```

Comment:
Having to specify the package identifier string prefixes manually feels a bit limited. Can we somehow remove all package identifiers without needing to know their names beforehand?

Reply:
Commented inline, I don't think it's a limitation. Explicit and simple is better in tests.